pkg/boot/boottest/json.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package boottest
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/u-root/u-root/pkg/boot"
+	"github.com/u-root/u-root/pkg/curl"
+)
+
+func module(r io.ReaderAt) map[string]interface{} {
+	m := make(map[string]interface{})
+	if f, ok := r.(curl.File); ok {
+		m["url"] = f.URL().String()
+	} else if f, ok := r.(fmt.Stringer); ok {
+		m["stringer"] = f.String()
+	}
+	return m
+}
+
+// CompareImagesToJSON compares the names, cmdlines, and file URLs in imgs to
+// the ones stored in jsonEncoded.
+func CompareImagesToJSON(imgs []boot.OSImage, jsonEncoded []byte) error {
+	var want interface{}
+	if err := json.Unmarshal(jsonEncoded, &want); err != nil {
+		return fmt.Errorf("failed to unmarshall test json %q: %v", jsonEncoded, err)
+	}
+
+	got := ImagesToJSONLike(imgs)
+	if !cmp.Equal(want, got) {
+		return fmt.Errorf("mismatch(-want, +got):\n%s", cmp.Diff(want, got))
+	}
+	return nil
+}
+
+// ToJSONFile can be used to generate JSON-comparable files for use with
+// CompareImagesToJSON in tests.
+func ToJSONFile(imgs []boot.OSImage, filename string) error {
+	enc, err := json.MarshalIndent(ImagesToJSONLike(imgs), "", "  ")
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, enc, 0644)
+}
+
+// ImagesToJSONLike spits out a json-convertible reproducible representation of
```

Comment:
This feels like a lot of overhead. Why not call MarshalIndent directly on boot.OSImage?

Reply:
1. ReaderAt's don't serialize into json nicely.

2. It also adds the image_type. You wouldn't be able to distinguish a linux and multiboot image if they both only had the Name, Kernel, and Cmdline fields present.

Compare:

```
2020/06/03 22:05:54 bare image in json:                   
{                                                                                                                   
  "Name": "Test this ^media \u0026 start Fedora-Workstation-Live 27",                                               
  "Kernel": {                                                                                                       
    "ReaderAt": {}                                        
  },                                                                                                                
  "Initrd": {                                                                                                       
    "ReaderAt": {}                                                                                                  
  },                                             
  "Cmdline": "initrd=initrd.img root=live:CDLABEL=Fedora-WS-Live-27-1-6 rd.live.image rd.live.check quiet"          
}                                                                                                                   
2020/06/03 22:05:54 our json:                                                                                                                                                                                                           
{                                                                                                                   
  "cmdline": "initrd=initrd.img root=live:CDLABEL=Fedora-WS-Live-27-1-6 rd.live.image rd.live.check quiet",         
  "image_type": "linux",                                                                                            
  "initrd": {                                      
    "url": "file://testdata/fedora-27-install/isolinux/initrd.img"                                                                                                                                                                      
  },                                                                                                                
  "kernel": {                                                                                                       
    "url": "file://testdata/fedora-27-install/isolinux/vmlinuz"                                                     
  },                                                                                                                
  "name": "Test this ^media \u0026 start Fedora-Workstation-Live 27"                                                
}                             
```